File: analysis/path_constraints.py
# Ethan Doughty
# path_constraints.py
"""Branch-aware path constraint stack for witness generation.

Tracks which branch conditions were active at each warning site,
enabling witnesses to reflect the control-flow context.
"""

from __future__ import annotations
from ir import Expr, BinOp, Var, Const, Neg


def _format_condition_expr(expr: Expr) -> str:
    """Format an IR condition expression as a human-readable string.

    Handles simple comparisons like BinOp(op='>', left=Var('n'), right=Const(3))
    -> 'n > 3'.  Falls back to 'condition at line N' for complex expressions.
    """
    if isinstance(expr, BinOp) and expr.op in {">", ">=", "<", "<=", "==", "~="}:
        left_str = _format_simple(expr.left)
        right_str = _format_simple(expr.right)
        if left_str is not None and right_str is not None:
            return f"{left_str} {expr.op} {right_str}"
    return f"condition at line {expr.line}"


def _format_simple(expr: Expr):
    """Return string for simple atoms (Var, Const, Neg(Const)), else None."""
    if isinstance(expr, Var):
        return expr.name
    if isinstance(expr, Const):
        v = expr.value
        if v == int(v):
            return str(int(v))
        return str(v)
    if isinstance(expr, Neg) and isinstance(expr.operand, Const):
        v = expr.operand.value
        if v == int(v):
            return str(-int(v))
        return str(-v)
    return None


class PathConstraintStack:
    """Stack of active branch conditions at the current analysis point.

    Each entry is (condition_description: str, branch_taken: bool, line: int).
    """

    def __init__(self):
        self._stack: list = []

    def push(self, condition_expr: Expr, branch_taken: bool, line: int) -> None:
        """Push a branch condition onto the stack."""
        desc = _format_condition_expr(condition_expr)
        self._stack.append((desc, branch_taken, line))

    def pop(self) -> None:
        """Pop the most recent branch condition. Defensive: no-op on empty stack."""
        if self._stack:
            self._stack.pop()

    def snapshot(self) -> tuple:
        """Return a frozen snapshot of the current stack."""
        return tuple(self._stack)

File: src/PathConstraints.fs
module PathConstraints

open Ir

// ---------------------------------------------------------------------------
// Condition formatting helpers (mirrors path_constraints.py)
// ---------------------------------------------------------------------------

let private formatSimple (expr: Expr) : string option =
    match expr with
    | Var(_, name)  -> Some name
    | Const(_, v)   ->
        if v = System.Math.Floor(v : float) then Some (string (int64 v))
        else Some (string v)
    | Neg(_, Const(_, v)) ->
        if v = System.Math.Floor(v : float) then Some (string -(int64 v))
        else Some (string -v)
    | _ -> None

let private formatConditionExpr (expr: Expr) : string =
    match expr with
    | BinOp(_, op, left, right)
        when Set.contains op (Set.ofList [">"; ">="; "<"; "<="; "=="; "~="]) ->
        match formatSimple left, formatSimple right with
        | Some l, Some r -> l + " " + op + " " + r
        | _              -> "condition at line " + string expr.Line
    | _ -> "condition at line " + string expr.Line

// ---------------------------------------------------------------------------
// PathConstraintStack: branch condition tracking
// Each entry: (condition_description, branch_taken, line)
// ---------------------------------------------------------------------------

type PathConstraintEntry = {
    description: string
    branchTaken: bool
    line:        int
}

type PathConstraintStack() =
    let mutable _stack : PathConstraintEntry list = []

    member _.Push(conditionExpr: Expr, branchTaken: bool, line: int) =
        let desc = formatConditionExpr conditionExpr
        _stack <- { description = desc; branchTaken = branchTaken; line = line } :: _stack

    member _.Pop() =
        match _stack with
        | _ :: rest -> _stack <- rest
        | []        -> ()  // defensive no-op

    member _.Snapshot() : PathConstraintEntry list =
        List.rev _stack   // oldest first, matching Python tuple order

    member _.Stack = _stack
